packages/compass-connections/src/connections-manager.ts

```diff
@@ -0,0 +1,223 @@
+  ConnectionAttemptSuccessful = 'connection-attempt-successful',
+  ConnectionAttemptFailed = 'connection-attempt-failed',
+  ConnectionDisconnected = 'connection-disconnected',
+}
+
+type ConnectionManagerEventListeners = {
+  [ConnectionsManagerEvents.ConnectionAttemptStarted]: (
+    connectionInfoId: ConnectionInfoId
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionAttemptCancelled]: (
+    connectionInfoId: ConnectionInfoId
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionAttemptSuccessful]: (
+    connectionInfoId: ConnectionInfoId,
+    dataService: DataService
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionAttemptFailed]: (
+    connectionInfoId: ConnectionInfoId,
+    error: Error
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionDisconnected]: (
+    connectionInfoId: ConnectionInfoId
+  ) => void;
+};
+
+export enum ConnectionStatus {
+  Disconnected = 'disconnected',
+  Connecting = 'connecting',
+  Connected = 'connected',
+  Failed = 'failed',
+}
+
+const { log } = createLoggerAndTelemetry('CONNECTIONS-MANAGER');
+
+export class ConnectionsManager extends EventEmitter {
+  private connectionAttempts = new Map<ConnectionInfoId, ConnectionAttempt>();
+  private connectionStatuses = new Map<ConnectionInfoId, ConnectionStatus>();
+  private dataServices = new Map<ConnectionInfoId, DataService>();
+
+  constructor(private readonly __TEST_CONNECT_FN?: ConnectFn) {
+    super();
+  }
+
+  getDataServiceForConnection(
+    connectionInfoId: ConnectionInfoId
+  ): DataService | undefined {
+    return this.dataServices.get(connectionInfoId);
+  }
+
+  statusOf(connectionInfoId: ConnectionInfoId): ConnectionStatus {
+    return (
+      this.connectionStatuses.get(connectionInfoId) ??
+      ConnectionStatus.Disconnected
+    );
+  }
+
+  cancelAllConnectionAttempt(): void {
+    for (const connectionInfoId of this.connectionAttempts.keys()) {
+      this.cancelConnectionAttempt(connectionInfoId);
+    }
+  }
+
+  cancelConnectionAttempt(connectionInfoId: ConnectionInfoId): void {
+    const connectionAttempt = this.connectionAttempts.get(connectionInfoId);
+    connectionAttempt?.cancelConnectionAttempt();
+    this.connectionAttempts.delete(connectionInfoId);
+    this.notifyConnectionAttemptCancelled(connectionInfoId);
+  }
+
+  /**
+   * @param connectionInfo The adjusted ConnectionInfo object that already has
+   * parameters such as appName.
+   */
+  async connect(connectionInfo: ConnectionInfo): Promise<void> {
+    try {
+      if (
+        [ConnectionStatus.Connecting, ConnectionStatus.Connected].includes(
+          this.statusOf(connectionInfo.id)
+        )
+      ) {
+        return;
+      }
+
+      const connectionAttempt = createConnectionAttempt({
+        logger: log.unbound,
+        connectFn: this.__TEST_CONNECT_FN,
+      });
+      this.connectionAttempts.set(connectionInfo.id, connectionAttempt);
+
+      this.notifyConnectionAttemptStarted(connectionInfo.id);
```

Comment:
Should the order be reversed here? If starting connection attempt fails syncronously, you'll get a failed notification before anything was even started (not sure if that's by design or not, but feels a bit off)

Reply:
Yea this makes sense. I had it like this because I wanted the notification to signify an actual start of connection attempt hence closer to `connect` but then I am missing this case so I will adjust, thank you :D